std/experimental/allocator/building_blocks/ascending_page_allocator.d

```diff
@@ -0,0 +1,418 @@
+module std.experimental.allocator.building_blocks.ascending_page_allocator;
+import std.experimental.allocator.common;
+/**
+`AscendingPageAllocator` is a fast and safe allocator which rounds all allocations
+to page size multiples. It reserves a range of virtual addresses (`mmap` for Posix
+and `VirtualAlloc` for Windows) and allocates memory at consecutive virtual
+addresses.
+
+When a chunk of memory is requested, the allocator finds a range of
+virtual pages which satisfy the requested size, changing their protection to
+read and write using OS primitives (`mprotect` and `VirtualProtect`).
+The physical memory is allocated on demand, when the pages are accessed.
+
+Deallocation removes any read/write permissions from the target pages
+and notifies the OS to reclaim the physical memory, while keeping the virtual
+memory.
```

Comment:
The thought that I just had is that you don't need to mess with the protection at all. If you just allocate from a pool and give back the arrays, you can rely on the bounds check to protect you from going out of bounds. Because, if you have two buffers adjacently allocated in your address range, there's really no protection at all (you don't leave range of memory with no access between them), and writing over the end of the first one will corrupt memory. So, protecting the memory by the OS only works in some cases, and not all of them.

Reply:
Yes, changing protection to none does not solve this alone. Maybe I can add guard pages between allocations for the ultimate defense :)

Reply:
@andralex @jercaianu all right, thanks, with that note in mind both this and the no-holes-reuse makes sense. With re-reading the documentation I can see the idea, but maybe it needs to be articulated more clearer to the casual observer.